spatial/curve/hilbert_test.go

```diff
@@ -0,0 +1,381 @@
+					v := make([]int, N)
+					for i := range v {
+						v[i] = rand.Intn(1 << ord)
+					}
+					u := make([]int, N)
+					for n := 0; n < b.N; n++ {
+						copy(u, v)
+						h.Curve(u)
+					}
+				})
+			}
+		})
+
+		b.Run(fmt.Sprintf("%dD/Space", N), func(b *testing.B) {
+			for ord := 1; ord <= 10; ord++ {
+				b.Run(fmt.Sprintf("Order %d", ord), func(b *testing.B) {
+					h := newCurve(ord, N)
+					d := rand.Intn(1 << (ord * N))
+					for n := 0; n < b.N; n++ {
+						h.Space(d)
+					}
+				})
+			}
+		})
+	}
+}
+
+type curve interface {
+	Dims() []int
+	Len() int
+	Curve(v []int) int
+	Space(d int) []int
+}
+
+func newCurve(order, dim int) curve {
+	switch dim {
+	case 2:
+		return Hilbert2D{Order: order}
+	case 3:
+		return Hilbert3D{Order: order}
+	case 4:
+		return Hilbert4D{Order: order}
+	}
+	panic("invalid dimension")
+}
+
+func testCurve(t *testing.T, c curve) {
+	t.Helper()
+
+	var errc int
+	fail := func() {
+		if errc < 10 {
+			errc++
+			t.Fail()
+		} else {
+			t.FailNow()
+		}
+	}
+
+	m := map[int][]int{}
+	curveRange(c, func(v []int) {
+		d := c.Curve(dup(v))
+		u := c.Space(d)
+		if !reflect.DeepEqual(v, u) {
+			t.Logf("Space is not the inverse of Curve for d=%d %v", d, v)
+			fail()
+		}
+
+		m[d] = dup(v)
+	})
+
+	D := 1
+	for _, v := range c.Dims() {
+		D *= v
+	}
+	for d := 0; d < D-1; d++ {
+		v, u := m[d], m[d+1]
+		if !adjacent(v, u) {
+			t.Logf("points %x and %x are not adjacent", d, d+1)
+			t.Logf("    %v -> %v", v, u)
+			fail()
+		}
+	}
+}
+
+func curveRange(c curve, fn func([]int)) {
+	size := c.Dims()
+	dimRange(len(size), size, make([]int, len(size)), fn)
+}
+
+func dimRange(dim int, size []int, v []int, fn func([]int)) {
+	if dim == 0 {
+		fn(v)
+		return
+	}
+
+	for i := 0; i < size[dim-1]; i++ {
+		v[dim-1] = i
+		dimRange(dim-1, size, v, fn)
+	}
+}
+
+// v may get mangled by Curve, so copy it.
+func dup(v []int) []int {
+	u := make([]int, len(v))
+	copy(u, v)
+	return u
+}
+
+// adjacent returns true if the euclidean distance between v and u is
+// exactly one. v and u must be the same length.
+//
+// In other words, given d(i) = abs(v[i] - u[i]), adjacent returns false if
+// d(i) > 1 for any i or if d(i) == 1 for more than a single i, or if d(i)
+// == 0 for all i.
+func adjacent(v, u []int) bool {
+	n := 0
+	for i := range v {
+		x := v[i] - u[i]
+		if x == 0 {
+			continue
+		}
+		if x < -1 || 1 < x {
+			return false
+		}
+		n++
+	}
+
+	return n == 1
+}
+
+func testCurveCase(t *testing.T, c curve, order int, want []int) {
+	t.Helper()
+
+	dim := len(c.Dims())
+	got := make([]int, len(want))
+	for i := range want {
+		v := vec(i, order, dim)
+		got[i] = c.Curve(dup(v))
+	}
+	if !reflect.DeepEqual(want, got) {
+		t.Fatalf("unexpected result:\ngot:  %v\nwant: %v", got, want)
+	}
+
+	for i, d := range want {
+		v := vec(i, order, dim)
+		if !reflect.DeepEqual(v, c.Space(d)) {
+			t.Fatalf("[%d] expected %v for d = %d", i, v, d)
+		}
+	}
+}
+
+func vec(n, ord, dim int) []int {
```

Comment:
On the other hand, this could have a comment what it does.

Reply:
First I have to remember what it does :facepalm: